packages/belos/tpetra/src/BelosTpetraOperator.hpp

```diff
@@ -0,0 +1,200 @@
+class TpetraOperator : public virtual Tpetra::Operator<Scalar, LocalOrdinal, GlobalOrdinal, Node> {
+public:
+  //! @name Typedefs
+  //@{
+  //! The Tpetra Multivector 
+  typedef Tpetra::MultiVector<Scalar, LocalOrdinal, GlobalOrdinal, Node> Tpetra_MultiVector;
+  //! The Tpetra Operator
+  typedef Tpetra::Operator<Scalar, LocalOrdinal, GlobalOrdinal, Node> Tpetra_Operator;
+  //! The Tpetra Map
+  typedef Tpetra::Map<LocalOrdinal,GlobalOrdinal,Node> Tpetra_Map;
+
+  //! @name Constructor / Destructor
+  //@{ 
+  
+  //! Constructor
+  TpetraOperator( const Teuchos::RCP<LinearProblem<Scalar,Tpetra_MultiVector,Tpetra_Operator> >& lp, 
+		  const Teuchos::RCP<Teuchos::ParameterList>& plist,
+                  bool initSolnVec = false )  
+  : lp_(lp), 
+    plist_(plist),
+    initSolnVec_(initSolnVec)
+  {
+  // Get the solver's name from the parameter list, use block Gmres by default.
+  std::string solver = plist_->get("Solver", "BlockGmres");
+
+  // Create a label for this Tpetra_Operator.
+  std::string solver_name = "Belos " + solver + " Solver";
+ 
+  // Copy std::string to character array.  
+  // Not using conversion routine copy() because it's not supported by RW on Janus. (HKT 11/13/2003) 
+  Solver.resize(solver_name.length()+1);
+  for (int i=0; i<(int)solver_name.length(); i++) {
+    Solver[i] = solver_name[i];
+  }   
+  Solver[solver_name.length()] = 0;
+
+  //  
+  // Create solver and solve problem.  This is inefficient, an instance of the solver should
+  // exist already and just be reset with a new RHS.
+  //  
+  if (solver == "BlockGmres") {
+    solver = "Block Gmres";
+  }   
+  else if (solver == "PseudoBlockGmres") {
+    solver = "Pseudo Block Gmres";
+  }
+  else if (solver == "BlockCG") {
+    solver = "Block CG";
+  }
+  else if (solver == "PseudoBlockCG") {
+    solver = "Pseudo Block CG";
+  }
+
+  plist_->remove( "Solver" );
+  
+  Belos::SolverFactory<Scalar,Tpetra_MultiVector,Tpetra_Operator> factory;
+  solver_ = factory.create( solver, plist );
+  solver_->setProblem( lp_ );
+
+}
+
+  
+  //! Destructor
+  virtual ~TpetraOperator() {}
+  //@}
+  
+  //! @name Operator application methods
+  //@{ 
+  
+  //! Apply the operator.
+  void apply( const Tpetra_MultiVector &X, Tpetra_MultiVector &Y, 
+           Teuchos::ETransp mode = Teuchos::NO_TRANS,
+           Scalar alpha = Teuchos::ScalarTraits<Scalar>::one(),
+           Scalar beta = Teuchos::ScalarTraits<Scalar>::zero()) const
+  {
+    Teuchos::RCP<const Tpetra_MultiVector> vec_X = Teuchos::rcp( &X, false );
+    Teuchos::RCP<Tpetra_MultiVector> vec_Y = Teuchos::rcp( &Y, false );
+    if (initSolnVec_)
+    {
+      vec_Y->putScalar( 0.0 );
+      lp_->setInitResVec( vec_X );
+    }
+    lp_->setProblem( vec_Y, vec_X );
+    Belos::ReturnType ret = solver_->solve();
+  };
+  //@}
+  
+  //! @name Attribute access functions
+  //@{ 
+
+  //! Return the label of the operator.
+  const char* Label() const { return(&Solver[0]); };
```

Comment:
1. `return Solver.data();`.  We have C++11 now.  If the `std::vector` is empty, `&Solver[0]` could throw in `operator[]` (so-called "checked" STL implementations can and do).
2. Method names should be lower case.
3. Do you mean to override a `label()` method in a base class?  If so, then I don't think the solver name is really the right "label."  A "label" is something a user can specify, and it can be unique, even for two objects that are otherwise the same.  If you really need this method (ask yourself first why you do), then consider renaming it to something like "solverName".

Reply:
This does not override of a base class function.  Changed the name for clarity.